### PR TITLE
[BREAKING CHANGE] Hardcoding a Default `assume_valid_target` to Reduce the Timecost of Block Synchronization in IBD mode

### LINKED ISSUE #4253: Hardcoding a Default `assume_valid_target` to Reduce the Timecost of Block Synchronization
#### Is your feature request related to a problem? Please describe.
We are proposing the hardcoding of a default `assume_valid_target` in order to reduce the timecost of CKB's block synchronization.
#### Describe the solution you'd like

The desired solution involves implementing:
- A default `assume_valid_target` for both the `mainnet` and `testnet`. The hardcoded default "target" block should be from `60 days` ago.
- If `assume_valid_target` is `0x0000000000000000000000000000000000000000000000000000000000000000`, then CKB will disable the default behavior and execute full verification for all blocks/
- The hardcoded value should be updated prior to every CKB's release.

image

### PR DESCRIPTION
**Pull request by eval-exec**
### What problem does this PR solve?

Issue Number: close #4253 

What's Changed:

### Related changes

- Set a block in 60 days ago as default assume_valid_target if user doesn't explictly specify `--assume-valid-target` argument.
- User specified `--assume-valid-target` will overwrite the default hardcoded `assume_valid_target`
- If user specify `--assume-valid-target 0x0000000000000000000000000000000000000000000000000000000000000000`, ckb will do full verification.

### full block synchronization result
image
block synchronize from `0` to `12,500,000`,
v0.115.0-rc2: need `30h40m`
this PR: need `12h40m`

This PR's performance test results show a `59%` reduction in block synchronization time compared to `v0.115.0-rc2`.

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code ci-runs-only: [ quick_checks,linters ]

Side effects

- None

### Release note 

```release-note
Title Only: Include only the PR title in the release note.
```

### CODE REVIEW

**Comment by eval-exec** [comment]
In the early phase of "finding assume_valid_target" CKB will print many logs like:
```
32(0xe9b97767424dd04aa65a1f7ad562b0faf8dd0fbf2a213d1586ea7969160f5996) please wait
2023-12-04 07:51:20.256 +00:00 BlockDownload INFO ckb_sync::synchronizer  best known header number: 10890000, hash: Byte32(0x662ba941ac8dd506832704a35fc13534e1ac402aa708670936af03d27114c643), can't find assume valid target temporarily, hash: Byte32(0xe9b97767424dd04aa65a1f7ad562b0faf8dd0fbf2a213d1586ea7969160f5996) please wait
2023-12-04 07:51:22.676 +00:00 BlockDownload INFO ckb_sync::synchronizer  best known header number: 10900000, hash: Byte32(0xc793a8184a272be18a0736b4a3713f19f75ba9f8913c88ca8d65b05a03304306), can't find assume valid target temporarily, hash: Byte32(0xe9b97767424dd04aa65a1f7ad562b0faf8dd0fbf2a213d1586ea7969160f5996) please wait
2023-12-04 07:51:25.162 +00:00 BlockDownload INFO ckb_sync::synchronizer  best known header number: 10910000, hash: Byte32(0xa7d0e834174d4c0677648c66b8d74868ce4dcf8514d8f7c40c909cf6f68578ff), can't find assume valid target temporarily, hash: Byte32(0xe9b97767424dd04aa65a1f7ad562b0faf8dd0fbf2a213d1586ea7969160f5996) please wait
2023-12-04 07:51:27.096 +00:00 BlockDownload INFO ckb_sync::synchronizer  best known header number: 10920000, hash: Byte32(0xe4162e3db5b73eedffa5267595f4b6596cf765b5b6afeae7b1e1649eff057eed), can't find assume valid target temporarily, hash: Byte32(0xe9b97767424dd04aa65a1f7ad562b0faf8dd0fbf2a213d1586ea7969160f5996) please wait
2023-12-04 07:51:29.217 +00:00 BlockDownload INFO ckb_sync::synchronizer  best known header number: 10930000, hash: Byte32(0x43ce8a7e8c1d65f313d8d1b217145477d5cbc24e12a5ffc9104da2e7ab0d0a04), can't find assume valid target temporarily, hash: Byte32(0xe9b97767424dd04aa65a1f7ad562b0faf8dd0fbf2a213d1586ea7969160f5996) please wait
2023-12-04 07:51:31.193 +00:00 BlockDownload INFO ckb_sync::synchronizer  best known header number: 10940000, hash: Byte32(0x418fe666fa6c47b98f79c2f6d579b590e4c3449dc555614063f4d3450df32852), can't find assume valid target temporarily, hash: Byte32(0xe9b97767424dd04aa65a1f7ad562b0faf8dd0fbf2a213d1586ea7969160f5996) please wait
2023-12-04 07:51:33.148 +00:00 BlockDownload INFO ckb_sync::synchronizer  best known header number: 10950000, hash: Byte32(0xa17857353561c5488bb0d82f4ad60114c96a3032eb9cbca771fc40e43ecb51c0), can't find assume valid target temporarily, hash: Byte32(0xe9b97767424dd04aa65a1f7ad562b0faf8dd0fbf2a213d1586ea7969160f5996) please wait
```
These logs will appear about 1 hour, need to reword these into more friendly message, like:
```
INFO current best known header: {number}-{hash},\
 assume_valid_target: {number}-{hash},\
 CKB need {duration} to reach the target, \
progress {27%}
```

File: devtools/release/find_block_60_days_ago.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env bash
```
Comment (doitian) [inline devtools/release/find_block_60_days_ago.sh:1]: Could you let this script update the file [util/constant/src/sync.rs](https://github.com/nervosnetwork/ckb/pull/4254/files#diff-de161884740a17e0b012a104898c602a2284c34f3d33eca7b76fe6602047b8fb) directly?

File: devtools/release/update_default_valid_target.sh [hunk 2]
```diff
@@ -43,17 +43,29 @@ function get_60days_ago_block(){
...
   ASSUME_TARGET_TIMESTAMP=$(get_block_timestamp ${host} ${ASSUME_TARGET_HEIGHT})
   ASSUME_TARGET_DATE=$(date -d @$((${ASSUME_TARGET_TIMESTAMP} / 1000)))
-  printf "the 60 days ago block is: %d %s in %s\n" ${ASSUME_TARGET_HEIGHT} ${ASSUME_TARGET_HASH} "${ASSUME_TARGET_DATE}"
+  EXPLORER_URL=${explorer_url}/block/${ASSUME_TARGET_HASH}
+  printf "the 60 days ago block is: %d %s in %s\n" ${ASSUME_TARGET_HEIGHT_DECIMAL} ${ASSUME_TARGET_HASH} "${ASSUME_TARGET_DATE}"
+  printf "you can view this block in ${EXPLORER_URL}\n\n"
+
+  TEXT="    // Default assume valid target for ${network}, expect to be a block 60 days ago.\n    // Need to update when CKB's new release\n    // in ${network}: the 60 days ago block is:\n    // height: ${ASSUME_TARGET_HEIGHT_DECIMAL}\n    // hash: ${ASSUME_TARGET_HASH}\n    // date: ${ASSUME_TARGET_DATE}\n    // you can view this block in ${EXPLORER_URL}\n    pub const DEFAULT_ASSUME_VALID_TARGET: &str =\n        \"${ASSUME_TARGET_HASH}\";"
+
+sed -i "/pub mod ${network} {/,/}/c\pub mod ${network} {\n${TEXT}\n}" util/constant/src/default_assume_valid_target.rs
```
Comment (doitian) [inline devtools/release/update_default_valid_target.sh:61]: `sed -i` does not work for freebsd sed used by macOS.

Use `sed -i.bak ` then remove the file `util/constant/src/default_assume_valid_target.rs.bak`

```suggestion
  sed -i.bak "/pub mod ${network} {/,/}/c\pub mod ${network} {\n${TEXT}\n}" util/constant/src/default_assume_valid_target.rs
  rm -f util/constant/src/default_assume_valid_target.rs.bak
```

**Review by doitian — CHANGES_REQUESTED** [review]

File: devtools/release/update_default_valid_target.sh (hunk 2 above)
Comment (eval-exec) [reply devtools/release/update_default_valid_target.sh:61]: OK

**Comment by eval-exec** [comment]
@doitian How about creating a template file named `default_assume_valid_target.rs.template` and then using `envsubst` to generate `default_assume_valid_target.rs` from this template file?

Ref: https://www.baeldung.com/linux/envsubst-command

File: devtools/release/update_default_valid_target.sh [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  ASSUME_TARGET_HEIGHT=$(get_60days_ago_block ${host})
+  ASSUME_TARGET_HEIGHT_DECIMAL=$(printf "%d" ${ASSUME_TARGET_HEIGHT})
+  ASSUME_TARGET_HASH=$(get_block_hash ${host} ${ASSUME_TARGET_HEIGHT})
+  ASSUME_TARGET_TIMESTAMP=$(get_block_timestamp ${host} ${ASSUME_TARGET_HEIGHT})
+  ASSUME_TARGET_DATE=$(date -d @$((${ASSUME_TARGET_TIMESTAMP} / 1000)))
+  EXPLORER_URL=${explorer_url}/block/${ASSUME_TARGET_HASH}
+  printf "the 60 days ago block is: %d %s in %s\n" ${ASSUME_TARGET_HEIGHT_DECIMAL} ${ASSUME_TARGET_HASH} "${ASSUME_TARGET_DATE}"
+  printf "you can view this block in ${EXPLORER_URL}\n\n"
+
+  TEXT="    // Default assume valid target for ${network}, expect to be a block 60 days ago.\n    // Need to update when CKB's new release\n    // in ${network}: the 60 days ago block is:\n    // height: ${ASSUME_TARGET_HEIGHT_DECIMAL}\n    // hash: ${ASSUME_TARGET_HASH}\n    // date: ${ASSUME_TARGET_DATE}\n    // you can view this block in ${EXPLORER_URL}\n    pub const DEFAULT_ASSUME_VALID_TARGET: &str =\n        \"${ASSUME_TARGET_HASH}\";"
```
Comment (eval-exec) [inline devtools/release/update_default_valid_target.sh:59]: The `TEXT` variable contains numerous spaces and newline characters, making it difficult to maintain. Using a template generation tool like `envsubst` can simplify our maintenance process.

**Comment by doitian** [comment]
> @doitian How about creating a template file named `default_assume_valid_target.rs.template` and then using `envsubst` to generate `default_assume_valid_target.rs` from this template file?
> 
> Ref: https://www.baeldung.com/linux/envsubst-command

I think you can just use shell HEREDOC string

```
cat <<EOF
/// The `mod mainnet` and `mod testnet`'s codes are generated
/// by script: ./devtools/release/update_default_valid_target.sh

/// sync config related to mainnet
pub mod mainnet {
    // Default assume valid target for mainnet, expect to be a block 60 days ago.
    // Need to update when CKB's new release
    // in mainnet: the 60 days ago block is:
    // height: $MAINNET_HEIGHT
    // hash: $MAINNET_HASH
    // date: $GENERATED_AT
    // you can view this block in https://explorer.nervos.org/block/$MAINNET_HASH
    pub const DEFAULT_ASSUME_VALID_TARGET: &str =
        "$MAINNET_HASH";
}

/// sync config related to testnet
pub mod testnet {
    // Default assume valid target for testnet, expect to be a block 60 days ago.
    // Need to update when CKB's new release
    // in testnet: the 60 days ago block is:
    // height: $TESTNET_HEIGHT
    // hash: $TESTNET_HASH
    // date: $GENERATED_AT
    // you can view this block in https://pudge.explorer.nervos.org/block/$TESTNET_HASH
    pub const DEFAULT_ASSUME_VALID_TARGET: &str =
        "$TESTNET_HASH";
}
EOF
```

**Review by doitian — APPROVED** [review]

**Comment by doitian** [comment]
`missing documentation for a module`

**Comment by eval-exec** [comment]
https://stackoverflow.com/questions/1167746/how-to-assign-a-heredoc-value-to-a-variable-in-bash

**Comment by eval-exec** [comment]
I will upload the full block synchronization chart for this PR later today.

**Comment by eval-exec** [comment]
I will upload the comparison of v0.115.0-rc2 and the full block synchronization chart from this PR several hours later.

**Comment by eval-exec** [comment]
> I will upload the comparison chart of the full block synchronization results between this PR and v0.115.0-rc2 in a few hours.

The comparison chart has been uploaded to the PR description.

I think this PR is ready to be reviewed.

**Comment by eval-exec** [comment]
> I will upload the comparison chart of the full block synchronization results between this PR and v0.115.0-rc2 in a few hours.

The comparison chart has been uploaded to the PR's description.

I think this PR is ready to be reviewed.

**Review by doitian — APPROVED** [review]

**Pull request merged**